istio/istio
# Webhook Warn on ineffective virtual service matches

Resolves https://github.com/istio/istio/issues/31525

cc @liamawhite 

To test, I used

```
apiVersion: networking.istio.io/v1alpha3
kind: VirtualService
metadata:
  name: issue31525
spec:
    http:
    - match:
      - uri:
           prefix: /
      route:
      - destination:
          host: httpbin
          port:
            number: 8000
    - match:
      - uri:
          prefix: /
      route:
      - destination:
          host: httpbin
          port:
            number: 8000
```

Note that you need K8s 1.19 to see the warnings from the API server.  The warnings now also show up in `istioctl validate`.

This PR just refactors the existing analyzer so that it isn't tied to the analysis framework, then has the validator call it.

- Moves some analysis code into a sub-package so it can be used as a library and avoid circular dependency
- Refactors the reporting of analysis messages out of the analyzing of duplicates.
- Adds logic to the existing validatator
